Fix/8209 Taxonomies Terms

MatteoPiovanelli opened the pull request:
fixes #8209

The way FullWeight used to be computed resulted in sometimes wrong values.
The FullWeight was not computed the same way in different places in the code.
The FullWeight ordering was not consistent with the ordering from 1.10.x.
The FullWeight ordering did not account for the Name property of TermParts.

This code is based off and targeted to the dev branch.

(I also took the chance to clean up and refactor some code that bothered me)

MatteoPiovanelli committed:
Refactored taxonomy service implementations as a first step to do the development for the fixes and upgrades.

MatteoPiovanelli committed:
Implemented new format for weight string.
Updated upgrade Process
working on drivers and handlers

MatteoPiovanelli committed:
changed service implementations to compute weights when needed. About to test

MatteoPiovanelli committed:
tested several operations related to taxonomies.

MatteoPiovanelli committed:
Fixed importing for terms

MatteoPiovanelli committed:
Fixed formula for FullWeight to account for negative weights
Added validation to value of Weight
Added a step to the migration to alter the FullWeight column and make it longer, so the system supports more than 15 nested levels of taxonomy (currently up to roughly 60).

MatteoPiovanelli committed:
updated cloning method

MatteoPiovanelli committed:
Fixed ordering by title of terms ot make it consistent with the old methods

MatteoPiovanelli committed:
fixed regression for draftable terms

sebastienros commented:
The upgrade code is from 1.10.x to dev?

@BenedekFarkas can you get it tested?

MatteoPiovanelli commented:
It should work to upgrade from 1.10 to dev as well as to upgrade dev to what is in this pr

MatteoPiovanelli commented:
Any news on this?
We've been running this on 40+ tenants for a few days, and so far we haven't had any issue with it. It's been boring to run the upgrade step on all of those, but apparently nothing broke yet.

BenedekFarkas commented:
Sorry @MatteoPiovanelli-Laser, reviewing/testing now!

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.Taxonomies/Drivers/TermPartDriver.cs:
```diff
@@ -94,23 +94,9 @@ public class TermPartDriver : ContentPartDriver<TermPart> {
-                containerTerm = containerTerm.Container.As<TermPart>();
-                fullWeightBuilder.Insert(0, containerTerm.Weight.ToString("D6") + "." + containerTerm.Id.ToString() + "/");
+            if (updater.TryUpdateModel(termPart, Prefix, null, null)) {
+                //termPart.FullWeight = _taxonomyService.ComputeFullWeight(termPart);
```
We can just call TryUpdateModel without the "if", because the inner code is commented out.

MatteoPiovanelli replied:
done

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.Taxonomies/Drivers/TermPartDriver.cs:
```diff
@@ -163,20 +149,23 @@ public class TermPartDriver : ContentPartDriver<TermPart> {
 
         protected override void Cloning(TermPart originalPart, TermPart clonePart, CloneContentContext context) {
-            clonePart.Count = originalPart.Count;
+            // arguably, copying the container/parent should be done elsewhere,
```
_A_rguably

MatteoPiovanelli replied:
done

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.Taxonomies/Handlers/TermPartHandler.cs:
```diff
@@ -17,6 +18,7 @@ public class TermPartHandler : ContentHandler {
             T = NullLocalizer.Instance;
             Filters.Add(StorageFilter.For(repository));
             OnInitializing<TermPart>((context, part) => part.Selectable = true);
+
```
Empty line should be after the closing braces.

MatteoPiovanelli replied:
done

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.Taxonomies/Migrations.cs:
```diff
@@ -108,5 +108,22 @@ public class Migrations : DataMigrationImpl {
+                table.CreateIndex("IDX_FullWeight", "FullWeight");
+            });
+            return 10;
+        }
```
UpdateFrom8 and UpdateFrom9 should be merged and also merged into Create (unfortunately, Create is stuck at 5 at the moment).

MatteoPiovanelli replied:
I think I split them apart like this because I needed to drop the index in a different method than the one where I altered the column.

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.Taxonomies/Services/TaxonomyService.cs:
```diff
@@ -30,7 +30,7 @@ public class TaxonomyService : ITaxonomyService {
         private readonly IShellDescriptorManager _shellDescriptorManager;
 
-        private readonly HashSet<int> _processedTermParts = new HashSet<int>(); 
+        private readonly HashSet<int> _processedTermParts = new HashSet<int>();
```
"_processedTermPartIds" shows better what it's used for.

MatteoPiovanelli replied:
I think I only changed the new line here, by reformatting the file, because I don't remember using this. I agree with you though

BenedekFarkas commented on src/Orchard.Web/Modules/Upgrade/Controllers/TaxonomyController.cs:
```diff
@@ -79,22 +79,26 @@ public class TaxonomyController : Controller {
             return RedirectToAction("Index");
         }
 
+        private const int BATCH = 50;
```
Should be "BatchSize". Btw class-level variables should be at the top before the ctor, so they are more visible.

BenedekFarkas reviewed (changes requested):
Did a quick code review and aside from a few minor details, it looks OK to me, but it's also a lot (can't really comment on ProcessPath and ComputeFullWeight).
Can you please show me some use-cases I should test?

MatteoPiovanelli commented:
I will look into the proposed changes. 
Let me know if there is any way I can also improve the comments around the code I added to make what's being done more clear.
Regarding examples:
The best way to test this, I think is:
 - Start from a 1.10.x tenant. Create a taxonomy there, whose terms are organized in a few levels, and with different weights. Some terms on a same level in the hierarchy should have the same weight to be ordered alphabetically.
 - "move" the tenancy to the current dev branch here. Run the upgrade process there to upgrade the FullWeight. In most cases the terms will not be ordered correctly. IN some cases something may actually break (I don't remember the exact conditions to break the current code off the top of my head see the issue #8209 for a bit more context).
 - merge the code form this PR in, run the upgrade again. The terms should be correctly ordered again.

As a side, the code as is in this PR gives a significant performance boost in the cases where a taxonomy is very large. We had a taxonomy with several hundred terms, and every time we tried to look at it opening the back office page would take several minutes. The FullWeight as computed in this PR allows proper ordering and pagination, so we don't have to do it in memory.

MatteoPiovanelli committed:
Changes asked in review.
I also changed the upgrade module so that the upgrade of the Weights is correctly in its own tab rather than in the tab for the upgrade that happened after orchard 1.7

MatteoPiovanelli committed:
Removed term upgrade from UI for older versions of Orchard.

MatteoPiovanelli commented:
I updated the PR. I think that answers the points raised in the review so far.
I also cleaned up the code in the upgrade module and separated this upgrade for the terms from the old one for Orchard 1.7.

sebastienros commented:
Waiting for @BenedekFarkas to approve the changes, thanks.

BenedekFarkas commented:
Yes, sorry, I'll take care of it hopefully this week.

sebastienros commented:
@BenedekFarkas

BenedekFarkas commented:
Sorry it took this long - tested with Matteo's instructions and looks good to me!
